[ts-sdk] Define the Provider Interface

666lcz opened the pull request:
For more context, please see https://github.com/MystenLabs/sui/pull/1308/files.

This PR defines the barebone interface for `Provider`. Subsequent PRs will add more methods and implement the method using a JSON-RPC client

pchrysochoidis commented on sdk/typescript/src/signers/signer.ts:
```diff
@@ -31,4 +36,61 @@ export abstract class Signer {
+  }
+
+  ///////////////////
+  // Sub-classes SHOULD leave these alone
```
these fns could be private right?

stella3d commented on sdk/typescript/src/signers/signer.ts:
```diff
@@ -31,4 +36,61 @@ export abstract class Signer {
+  ): Promise<TransactionResponse> {
+    this._checkProvider('signAndExecuteTransaction');
+    const sig = await this.signData(txBytes);
+    return await this.provider!.executeTransaction({
```
i've never seen this `provider!.execute()` syntax before - what is the `!` doing ?

666lcz replied:
That's the non-null assertion operator. It is a way to tell the compiler "this expression cannot be null or undefined here, so don't complain about the possibility of it being null or undefined." 

https://stackoverflow.com/questions/42273853/in-typescript-what-is-the-exclamation-mark-bang-operator-when-dereferenci

stella3d commented on sdk/typescript/src/signers/signer.ts:
```diff
@@ -31,4 +36,61 @@ export abstract class Signer {
+  ///////////////////
+  // Sub-classes SHOULD leave these alone
+
+  _checkProviderAndSerializer(operation?: string): void {
```
can we also eliminate all `_` prefixes here ?  an explicit typescript `private` modifier would communicate the intent more clearly

stella3d commented on sdk/typescript/src/signers/signer.ts:
```diff
@@ -31,4 +36,61 @@ export abstract class Signer {
+  async transfer(
+    transaction: TransferTransaction
+  ): Promise<TransactionResponse> {
+    this._checkProviderAndSerializer('transfer');
```
it seems like we're always checking for prerequisite objects before doing anything, but i'm not sure we gain any significant flexibility by allowing these objects to be null/undefined at all.  

Can we set the provider, serializer, & signer in the constructor, and make them private (they're already readonly) ? that way they're always defined.
If users want to change the signer, serializer, or provider, they can construct a new class instance.

666lcz replied:
Some use cases might just need a signer without a provider, e.g., to produce pre-signed message, to prove that the user owns this address, etc. This use case is very popular in some portfolio tracking applications such as https://debank.com/. 

My demo app also uses similar pattern: https://gist.github.com/666lcz/6cc6c983e841eda9f7a457e47bc473b7 without a provider.

stella3d replied:
actually, now that i think about this some more - I think it seems like we should completely remove the provider from the signer. Signing & sending the tx seem like two separate concerns. 

I think the ideal signer might be something that implements _**only**_ 
```ts
interface Signer { 
    signData(data: Base64DataBuffer): Promise<SignaturePubkeyPair>
}
```

stella3d commented on sdk/typescript/src/signers/signer.ts:
```diff
@@ -31,4 +36,61 @@ export abstract class Signer {
+
+  // Returns a new instance of the Signer, connected to provider.
+  // This MAY throw if changing providers is not supported.
+  abstract connect(provider: Provider): Signer;
```
i think we should remove this API in favor of having users construct a new instance

666lcz replied:
I thought about this and think this is pretty convenient in the case I mentioned above where the Signer only needs to be connected to a provider when it's necessary. If we do not expose this connect method, the caller needs to load the keypair again(because the keypair is private). Ethers.js has a similar interface: https://github.com/ethers-io/ethers.js/blob/948f77050dae884fe88932fd88af75560aac9d78/packages/abstract-signer/src.ts/index.ts#L81

stella3d replied:
OK, that makes sense - this actually does end up just calling a constructor, so I think it's ok, as long as it isn't changing an existing object

stella3d reviewed (changes requested):
I think we should change the way this is setup to eliminate the need for null/undefined checking in the signer (by making the constructor set the values).

We may also want to think through if there are advantages to exporting typescript `interface`s over abstract classes.

666lcz commented:
Addressed @stella3d 's comment by:
- defining a `Signer` interface that is `provider` free
- Rename the original `Signer` abstract class to `SignerWithProvider` to provide an ergonomic interface for the API caller(see example below)
- Added a `VoidProvider` so that we do not need to add checks to see if the provider is null

With these changes, we keep the `Signer` interface clean while also providing an ergonomic interface for the API caller.


The code snippets below will demonstrate why having a `SignerWithProvider` is more ergonomic than always having the `Signer` and the `Provider` separate

With the `SignerWithProvider` 
```
// NFT marketplace with non-custodial wallet
const provider = new sui.providers.WalletProvider(window.sui);
// for operations that does not need signing
const address = provider.getAddress();

// for operations that need signing
const signer = provider.getSigner();
signer.execute_move_call();
```

If we always have the `Provider` and `Signer` separate:
```
const provider = new sui.providers.WalletProvider(window.sui);
// for operations that does not need signing
const address = provider.getAddress();

// for operations that need signing
const data = provider.move_call_data();
const signer = provider.getSigner();
const signedData = signer.signData(data)
provider.execute(signedData)
```

stella3d reviewed (approved):
This latest update looks good, it addresses my previous concerns with initialization of dependency objects. great work
